Terminal.Gui/ConsoleDrivers/CursesDriver/CursesDriver.cs

```diff
@@ -1071,52 +1071,78 @@ public string GetClipboardData ()
 
 		public void SetClipboardData (string text)
 		{
-			var tempFileName = System.IO.Path.GetTempFileName ();
-			System.IO.File.WriteAllText (tempFileName, text);
+			// var tempFileName = System.IO.Path.GetTempFileName ();
+			// System.IO.File.WriteAllText (tempFileName, text);
+			// try {
+			// 	// BashRunner.Run ($"cat {tempFileName} | xsel -i --clipboard");
+			// 	BashRunner.Run ($"cat {tempFileName} | xclip -selection clipboard");
+			// } finally {
+			// 	System.IO.File.Delete (tempFileName);
+			// }
+
 			try {
-				// BashRunner.Run ($"cat {tempFileName} | xsel -i --clipboard");
-				BashRunner.Run ($"cat {tempFileName} | xclip -selection clipboard");
-			} finally {
-				System.IO.File.Delete (tempFileName);
+				BashRunner.Run ("xclip -selection clipboard -i", false, text);
+			} catch (Exception ex) {
+				throw new Exception (ex.Message);
```

Comment:
Should this be `NotSupportedException` so that there is a standard exception type across all the drivers? Either way you probably want to expose the ex as an InnerException e.g. `throw new NotSupportedException("Write to clipboard failed",ex);`

Reply:
Thanks. In your opinion what's the best solution?

Reply:
Not sure how feasible this is but I would try to extend `IClipboard` to the following (see below).  If it was possible to implement this then testing should become much easier as you can just return true from test if `IsSupported` is false.

```csharp
/// <summary>
/// Definition to interact with the OS clipboard.
/// </summary>
public interface IClipboard {

    /// <summary>
    /// Returns true if the environmental dependencies are in place to interact with the OS clipboard
    /// </summary>
    /// <value></value>
    bool IsSupported {get;}

    /// <summary>
    /// Sets the operation system clipboard.
    /// </summary>
    /// <param name="text"></param>
    /// <exception cref="NotSupportedException">Thrown if it was not possible to set the clipboard contents</exception>
    void SetClipboardData (string text);

    /// <summary>
    /// Sets the operation system clipboard if possible.
    /// </summary>
    /// <param name="text"></param>
    /// <returns>True if the clipboard content was set successfully</returns>
    bool TrySetClipboardData(string text);

    /// <summary>
    /// Gets the current contents of the OS clipboard
    /// </summary>
    /// <returns></returns>
    /// <exception cref="NotSupportedException">Thrown if it was not possible to read the clipboard contents</exception>
    string GetClipboardData ();

    /// <summary>
    /// Returns true if it was possible to read the OS clipboard
    /// </summary>
    /// <param name="result">Clipboard contents read</param>
    /// <returns></returns>
    bool TryGetClipboardData(out string result);
}
```

I think regardless of implementation `GetClipboardData` and `SetClipboardData` should only throw `NotSupportedException`.  You could enforce this rule with a shared abstract base implementation e.g.

```csharp
public abstract class ClipboardBase : IClipboard {
    public abstract bool IsSupported {get;}

    public string GetClipboardData ()
    {
        try
        {
            return GetClipboardDataImpl();
        }catch(Exception ex)
        {
            throw new NotSupportedException("Failed to read clipboard",ex);
        }
    }

    protected abstract string GetClipboardDataImpl ();

    public void SetClipboardData (string text)
    {	
        try
        {
            SetClipboardDataImpl(text);
        }catch(Exception ex)
        {
            throw new NotSupportedException("Failed to write to clipboard",ex);
        }
    }

    protected abstract void SetClipboardDataImpl (string text);

    public bool TryGetClipboardData (out string result)
    {
        // Don't even try to read because environment is not set up
        if(!IsSupported){
            result = null;
            return false;
        }


        try
        {
            result =  GetClipboardDataImpl();
            return true;
        }catch(Exception)
        {
            result = null;
            return false;
        }
    }

    public bool TrySetClipboardData (string text)
    {
        // Don't even try to set because environment is not set up
        if(!IsSupported){
            return false;
        }

        try
        {
            TrySetClipboardData(text);
            return true;
        }catch(Exception ex)
        {
            return false;
        }
    }
}
```

Reply:
This is challenging but I really like implementation. You know, if I find it difficult to implement it I'll ask you to implement :-). But I will try to put it into practice and still try a new implementation for the WSL, which is not worth installing the XCLIP because it will never work. I'm having some ideas and with your help, I think I'll get it right. Thanks.

Reply:
> Let me know if you think this is over engineering!

Well, it is not very easy :-). What should I use in each driver, the `IClipboard` or the `ClipboardBase`? I think there is no sense to use both. Can you please provide some link to learn more about this implementation. Thanks.

Reply:
> 
> 
> > Let me know if you think this is over engineering!
> 
> Well, it is not very easy :-). What should I use in each driver, the `IClipboard` or the `ClipboardBase`? I think there is no sense to use both. Can you please provide some link to learn more about this implementation. Thanks.

I would have `ConsoleDriver` still use `IClipboard` but make `WindowsClipboard`, `MacOSXClipboard` and `CursesClipboard` implement the abstract base.  Its usually better to reference the interface.  The abstract base is there to ensure consistency between implementations (e.g. the type of exception thrown and how to implement Try) as well as to cut out duplication between the clipboard implementations.

If you run into difficulties I can help out :)

Reply:
If something is wrong please let me know. Thanks @tznind.